src/v/kafka/server/group_metadata.cc

```diff
@@ -390,11 +392,31 @@ group_metadata_serializer make_backward_compatible_serializer() {
     };
     return group_metadata_serializer(std::make_unique<impl>());
 }
+iobuf maybe_unwrap_from_iobuf(iobuf buffer) {
+    /*
+     * Previously in redpanda we had an error leading to wrapping
+     * tombstone record keys with an iobuf serialization envelope. In order to
+     * provide compatibility and be able to read keys wrapped with an additional
+     * iobuf we introduce this check. The check validates the key version being
+     * in allowed range. otherwise we try to unwrap a key from iobuf.
+     *
+     * TODO: remove this code in future releases
+     *
+     */
+    iobuf_parser parser(buffer.copy());
+    int16_t key_version = request_reader(parser.peek(2)).read_int16();
+    if (unlikely(key_version < 0 || key_version > 2)) {
```

Comment:
>  The check validates the key version being in allowed range. otherwise we try to unwrap a key from iobuf.

the comment needs to explain how this works. i understand that there is a kafka version involved here, but with the iobuf wrapper, it's not exactly clear what the version is here. presumably the bytes overlap between instances of the iobuf wrapper length and the kafka version when no wrapper is present. but i haven't been able to deduce what is happening precisely.

Reply:
I've changed the approach a little bit here and added comment on why is the approach valid.